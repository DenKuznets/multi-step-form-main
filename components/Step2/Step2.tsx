'use client';

import PlanCard from './PlanCard/PlanCard';
import { Plans } from '../../utils/plans';
import { useRouter } from 'next/navigation';
import { useForm, SubmitHandler } from 'react-hook-form';
import { useAppDispatch, useAppSelector } from '@/lib/redux/hooks';
import {
    selectPaymentMethod,
    selectPlan,
    togglePaymentMethod,
    setPlan
} from '@/lib/redux/slices/appSlice';

export interface FormValues {
    plan: string;
}

export const PAYMENT = {
    MONTHLY: 'monthly',
    YEARLY: 'yearly'
};

const Step2 = () => {
    const router = useRouter();
    const dispatch = useAppDispatch();
    const currentPlan = useAppSelector(selectPlan);
    const currentPaymentMethod = useAppSelector(selectPaymentMethod);
    const { register, handleSubmit } = useForm<FormValues>({
        defaultValues: {
            plan: currentPlan
        }
    });

    const onSubmit: SubmitHandler<FormValues> = (data) => {
        dispatch(setPlan(data.plan));
        router.push('/add-ons');
    };

    const plansList = Plans.map((plan) => {
        return <PlanCard plan={plan} key={plan.name} register={register} />;
    });
    return (
        <>
            <form
                noValidate
                aria-label="select plan form"
                onSubmit={handleSubmit(onSubmit)}
            >
                <div className="flex flex-col gap-4 md:flex-row md:justify-evenly">
                    {plansList}
                </div>
                <div
                    data-testid="payment-switch"
                    className="mt-4 flex items-center justify-center gap-5 rounded-md bg-alabaster p-4 font-medium "
                >
                    <div
                        className={
                            currentPaymentMethod === PAYMENT.MONTHLY
                                ? 'text-marineBlue'
                                : 'text-coolGray'
                        }
                    >
                        Monthly
                    </div>
                    <div
                        data-testid="payment-switch-toggle"
                        onClick={() => {
                            dispatch(togglePaymentMethod());
                        }}
                        className={`relative flex h-5 w-10 cursor-pointer items-center rounded-full bg-marineBlue px-1 `}
                    >
                        <div
                            className={`relative h-3 w-3 rounded-full bg-white transition-all ${
                                currentPaymentMethod === PAYMENT.MONTHLY
                                    ? 'left-0'
                                    : 'left-full -translate-x-full'
                            }`}
                        ></div>
                    </div>
                    <div
                        className={
                            currentPaymentMethod === PAYMENT.YEARLY
                                ? 'text-marineBlue'
                                : 'text-coolGray'
                        }
                    >
                        Yearly
                    </div>
                </div>
                <div className="fixed bottom-0  left-0 flex w-full justify-between bg-white p-4 shadow sm:absolute sm:mt-8 sm:p-0 sm:shadow-none md:mb-4">
                    <button
                        onClick={(e) => {
                            e.preventDefault();
                            router.push('/personal-info');
                        }}
                        className="btn btn-back text-coolGray"
                    >
                        Go Back
                    </button>
                    <button className="btn btn-next ">next step</button>
                </div>
            </form>
        </>
    );
};

export default Step2;
