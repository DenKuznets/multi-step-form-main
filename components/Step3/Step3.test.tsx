import '@testing-library/jest-dom';
import {
    renderWithProviders,
    screen,
    within
} from '@/lib/redux/utils-for-tests';
import Step2 from './Step2';
import userEvent from '@testing-library/user-event';

jest.mock('next/navigation', () => ({
    useRouter() {
        return {
            prefetch: () => null
        };
    }
}));

test('renders correctly', () => {
    renderWithProviders(<Step2 />);

    const form = screen.getByRole('form', { name: /select plan form/i });
    const planCards = within(form).getAllByTestId('plan-card');
    const paymentSwitch = within(form).getByTestId('payment-switch');
    const backbutton = screen.getByRole('button', {
        name: /go back/i
    });
    const nextStepButton = screen.getByRole('button', {
        name: /next step/i
    });

    expect(form).toBeInTheDocument();
    expect(planCards).toHaveLength(3);
    expect(paymentSwitch).toBeInTheDocument();
    expect(backbutton).toBeInTheDocument();
    expect(nextStepButton).toBeInTheDocument();
});

test('renders Step2 unchanged', () => {
    const { container } = renderWithProviders(<Step2 />);
    expect(container).toMatchSnapshot();
});

test('pressing payment switch should change plan price from month to year and back', async () => {
    const user = userEvent.setup();
    renderWithProviders(<Step2 />);
    const arcadePlanPrice = screen.getByText(/\$9\/mo/i);
    expect(arcadePlanPrice).toHaveTextContent('$9/mo');
    const toggle = screen.getByTestId('payment-switch-toggle');
    await user.click(toggle);
    expect(arcadePlanPrice).toHaveTextContent('$90/yr');
    await user.click(toggle);
    expect(arcadePlanPrice).toHaveTextContent('$9/mo');
});
