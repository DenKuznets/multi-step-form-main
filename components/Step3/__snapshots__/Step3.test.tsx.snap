// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot Step3 unchanged 1`] = `
<div>
  <form
    aria-label="select add-ons form"
    novalidate=""
  >
    <div
      class="flex flex-col gap-4"
    >
      <div
        class="relative flex items-center rounded-md px-4 py-[11px] outline outline-1 outline-lightGray transition-all hover:outline-marineBlue lg:px-6 lg:py-5 [&:has(input:checked)]:bg-alabaster [&:has(input:checked)]:outline-marineBlue"
        data-testid="add-on"
      >
        <label
          class=" absolute left-0 top-0 z-10 h-full w-full cursor-pointer"
          for="online"
        />
        <input
          class="peer appearance-none checked:border-0"
          id="online"
          name="online"
          type="checkbox"
        />
        <div
          class="relative mr-4 h-5 w-5 rounded-[4px] border border-lightGray peer-checked:border-none [&>svg]:hidden peer-checked:[&>svg]:block"
        >
          <svg
            class="absolute left-0 top-0 h-full w-full"
            fill="none"
            height="20"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M2 4.5A2.5 2.5 0 0 1 4.5 2h15A2.5 2.5 0 0 1 22 4.5v15a2.5 2.5 0 0 1-2.5 2.5h-15A2.5 2.5 0 0 1 2 19.5v-15Zm16.787 5.075a1 1 0 0 0-.003-1.414l-.709-.705a1 1 0 0 0-1.414.003l-5.772 5.796L7.57 9.98a1 1 0 0 0-1.415.01l-.702.712a1 1 0 0 0 .01 1.414L10.2 16.79a1 1 0 0 0 1.411-.007l7.176-7.207Z"
              fill="hsl(243, 100%, 62%)"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="mr-auto"
        >
          <div
            class="text-sm font-bold text-marineBlue lg:text-base"
          >
            Online service
          </div>
          <div
            class="text-xs text-coolGray lg:text-sm lg:font-medium"
          >
            Access to multiplayer games
          </div>
        </div>
        <div
          class="text-xs font-bold text-blue-700 lg:text-sm lg:font-bold"
        >
          +$10/yr
        </div>
      </div>
      <div
        class="relative flex items-center rounded-md px-4 py-[11px] outline outline-1 outline-lightGray transition-all hover:outline-marineBlue lg:px-6 lg:py-5 [&:has(input:checked)]:bg-alabaster [&:has(input:checked)]:outline-marineBlue"
        data-testid="add-on"
      >
        <label
          class=" absolute left-0 top-0 z-10 h-full w-full cursor-pointer"
          for="storage"
        />
        <input
          checked=""
          class="peer appearance-none checked:border-0"
          id="storage"
          name="storage"
          type="checkbox"
        />
        <div
          class="relative mr-4 h-5 w-5 rounded-[4px] border border-lightGray peer-checked:border-none [&>svg]:hidden peer-checked:[&>svg]:block"
        >
          <svg
            class="absolute left-0 top-0 h-full w-full"
            fill="none"
            height="20"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M2 4.5A2.5 2.5 0 0 1 4.5 2h15A2.5 2.5 0 0 1 22 4.5v15a2.5 2.5 0 0 1-2.5 2.5h-15A2.5 2.5 0 0 1 2 19.5v-15Zm16.787 5.075a1 1 0 0 0-.003-1.414l-.709-.705a1 1 0 0 0-1.414.003l-5.772 5.796L7.57 9.98a1 1 0 0 0-1.415.01l-.702.712a1 1 0 0 0 .01 1.414L10.2 16.79a1 1 0 0 0 1.411-.007l7.176-7.207Z"
              fill="hsl(243, 100%, 62%)"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="mr-auto"
        >
          <div
            class="text-sm font-bold text-marineBlue lg:text-base"
          >
            Larger storage
          </div>
          <div
            class="text-xs text-coolGray lg:text-sm lg:font-medium"
          >
            Extra 1TB of cloud save
          </div>
        </div>
        <div
          class="text-xs font-bold text-blue-700 lg:text-sm lg:font-bold"
        >
          +$20/yr
        </div>
      </div>
      <div
        class="relative flex items-center rounded-md px-4 py-[11px] outline outline-1 outline-lightGray transition-all hover:outline-marineBlue lg:px-6 lg:py-5 [&:has(input:checked)]:bg-alabaster [&:has(input:checked)]:outline-marineBlue"
        data-testid="add-on"
      >
        <label
          class=" absolute left-0 top-0 z-10 h-full w-full cursor-pointer"
          for="customize"
        />
        <input
          checked=""
          class="peer appearance-none checked:border-0"
          id="customize"
          name="customize"
          type="checkbox"
        />
        <div
          class="relative mr-4 h-5 w-5 rounded-[4px] border border-lightGray peer-checked:border-none [&>svg]:hidden peer-checked:[&>svg]:block"
        >
          <svg
            class="absolute left-0 top-0 h-full w-full"
            fill="none"
            height="20"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M2 4.5A2.5 2.5 0 0 1 4.5 2h15A2.5 2.5 0 0 1 22 4.5v15a2.5 2.5 0 0 1-2.5 2.5h-15A2.5 2.5 0 0 1 2 19.5v-15Zm16.787 5.075a1 1 0 0 0-.003-1.414l-.709-.705a1 1 0 0 0-1.414.003l-5.772 5.796L7.57 9.98a1 1 0 0 0-1.415.01l-.702.712a1 1 0 0 0 .01 1.414L10.2 16.79a1 1 0 0 0 1.411-.007l7.176-7.207Z"
              fill="hsl(243, 100%, 62%)"
              fill-rule="evenodd"
            />
          </svg>
        </div>
        <div
          class="mr-auto"
        >
          <div
            class="text-sm font-bold text-marineBlue lg:text-base"
          >
            Customizable Profile
          </div>
          <div
            class="text-xs text-coolGray lg:text-sm lg:font-medium"
          >
            Custom theme on your profile
          </div>
        </div>
        <div
          class="text-xs font-bold text-blue-700 lg:text-sm lg:font-bold"
        >
          +$20/yr
        </div>
      </div>
    </div>
    <div
      class="fixed bottom-0  left-0 flex w-full justify-between bg-white p-4 shadow sm:absolute sm:mt-8 sm:p-0 sm:shadow-none md:mb-4"
    >
      <button
        class="btn btn-back text-coolGray"
      >
        Go Back
      </button>
      <button
        class="btn btn-next "
      >
        next step
      </button>
    </div>
  </form>
</div>
`;
